### PR TITLE
Implement TLS and enable it when a cert is available

### PR DESCRIPTION
**Pull request by Abuelodelanada**
## Issue

This PR implements TLS as part of this tracking issue https://github.com/canonical/cos-lite-bundle/issues/75

## How to test it

1. Deploy Loki
2. Deploy `self-signed-certificates`
3. Deploy `prometheus-k8s`
4. Relate Loki to `self-signed-certificates` and `prometheus-k8s`
5. You can check HTTP endpoint in Prometheus:
imagen

### CODE REVIEW

File: charmcraft.yaml
Comment (dstathis) [inline charmcraft.yaml:13]: We should not be doing this in production code. Out charms are supposed to be built from source.

(This is the blocking comment)

File: metadata.yaml
Comment (dstathis) [inline metadata.yaml:68]: I'm not sure I understand this comment.

File: src/config_builder.py
Comment (dstathis) [inline src/config_builder.py:90]: I do not think this is a safe way to check if TLS should be enabled. What if the relation exists but the cert has not been generated yet? In that case we don't want the server running over HTTP while we wait.

I am going to open a PR adding [this method](https://github.com/canonical/observability-libs/compare/main...enabled_method) to cert_manager. Until then, I think you should do something similar.

File: src/config_builder.py
Comment (dstathis) [inline src/config_builder.py:23]: Why add `*args` to the arguments? Do you expect anything else to be passed in?

**Review by dstathis — CHANGES_REQUESTED** [review]

File: charmcraft.yaml
Comment (Abuelodelanada) [reply charmcraft.yaml]: You are right, I forget this

File: src/config_builder.py
Comment (Abuelodelanada) [reply src/config_builder.py:90]: Sounds reasonable. 
That is a simple PR, let's wait for it to be merged, and then I'll use that method here.

File: src/charm.py
Comment (sed-i) [inline src/charm.py:265]: The new scrape job looks quite different:
- Iirc, there's a leader guard elsewhere surrounding this function call, so without the `*`-notation we'd only get this unit.
- If we do end up having ingress in place, we will have a path component and we will need to set `metrics_path` as before. Unless you refactor the charm to use `strip_prefix=True` in the ingress lib.
- `self.hostname` is `socket.getfqdn()`, so we're losing `self.ingress_per_unit.url` compared to before. When grafana agent pushes to loki. Perhaps add a comment that clearly states that we assume self-mon is always within the model?

File: src/config_builder.py
Comment (sed-i) [inline src/config_builder.py:23]: Should we replace it with c'tor args?
I'm inclined to having a dep-free config builder.

File: src/config_builder.py
Comment (sed-i) [inline src/config_builder.py:95]: Seems like this is mTLS.
Can't we omit them then?

File: src/config_builder.py
Comment (sed-i) [reply src/config_builder.py:90]: I'd prefer something like
```python
config = ConfigBuilder(tls_enabled=is_tls_enabled(), ...).build()
```
or
```python
config = ConfigBuilder(...).set_http_tls({"cert_path"=..., "key_path"=...}).build()
```

File: src/config_builder.py
Comment (Abuelodelanada) [reply src/config_builder.py:23]: c'tor args??

I also inclined to dep-free config builder.The problem here is that we need a lot of information from the charm to build the config, that why I'm injecting it.

File: src/charm.py
Comment (Abuelodelanada) [reply src/charm.py:265]: Until we have OB031 approved, let's keep self.ingress_per_unit.url

File: src/config_builder.py
Comment (dstathis) [reply src/config_builder.py:90]: Sounds good to me.

**Review by dstathis — APPROVED** [review]

File: src/charm.py
Comment (Abuelodelanada) [reply src/charm.py:265]: Yes, you are right.

So, we have another issue here then, because when the relation to Traefik is established (`ingress_per_unit`), we set this scrape job and we are always scrapping the leader unit: 


```python
        return [
            {
                "metrics_path": f"{external_url.path}/metrics",
                "static_configs": [{"targets": [f"{external_url.hostname}:{external_url.port}"]}],
            }
        ]
```

For instance, without Traefik we have:

imagen

imagen


With Traefik:

imagen

imagen

Comment (sed-i) [reply src/charm.py:265]: I imagine the peer data trick should do it.

File: src/charm.py
Comment (sed-i) [inline src/charm.py:293]: Now we have `replan` so we don't need to track the `restart`.

**Review by sed-i — APPROVED** [review]

**Pull request merged**
